src/Store/ExtensionClientStoreConnector.re

```diff
@@ -21,6 +21,60 @@ module Protocol = Extensions.ExtHostProtocol;
 
 module Workspace = Protocol.Workspace;
 
+module ExtensionCompletionProvider = {
+  let suggestionItemToCompletionItem:
+    Protocol.SuggestionItem.t => Model.CompletionItem.t =
+    suggestion => {
+      let completionKind =
+        suggestion.kind |> Option.bind(CompletionItemKind.ofInt);
+
+      {
+        label: suggestion.label,
+        kind: completionKind,
+        detail: suggestion.detail,
+      };
+    };
+
+  let suggestionsToCompletionItems:
+    option(Protocol.Suggestions.t) => list(Model.CompletionItem.t) =
+    fun
+    | Some(suggestions) =>
+      List.map(suggestionItemToCompletionItem, suggestions)
+    | None => [];
+
+  let create =
+      (
+        client: ExtHostClient.t,
+        suggestProvider: Protocol.SuggestProvider.t,
+        buffer,
+        _completionMeet,
+        position,
+      ) => {
+    Core.Buffer.getFileType(buffer)
+    |> Option.map(
+         Extensions.DocumentSelector.matches(suggestProvider.selector),
+       )
+    |> Option.bind(matches => {
+         let uri = Core.Buffer.getUri(buffer);
+         let position = Protocol.OneBasedPosition.ofPosition(position);
+
+         switch (matches) {
+         | false => None
+         | true =>
+           Some(
+             ExtHostClient.getCompletions(
+               suggestProvider.id,
+               uri,
+               position,
+               client,
+             )
+             |> Lwt.map(suggestionsToCompletionItems),
+           )
+         };
```

Comment:
`if` expressions still work fine in Reason (and isn't any less functional) ;) I've noticed a pattern with beginners especially tending to prefer pattern matching over `if` expressions. Probably just to feel it out, but `if` expressions tend to be better suited for boolean conditions because it's a less general construct, it puts the condition and "consequence" closer together, and it evolves more easily when more conditions are required.